[Woo POS] MVP Analytics: Card payment success event properties

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

## Description
This PR it's a continuation of https://github.com/woocommerce/woocommerce-ios/pull/15138, and adds the following missing properties to `pos_card_present_collect_payment_success` event:

```
milliseconds_since_order_creation_success
milliseconds_since_reader_ready_to_collect_payment
milliseconds_since_card_tapped
checkout_tap_count
```
In order to track these we follow the same design that the existing `milliseconds_since_customer_interaction_started` property, as we need to keep track of the elapsed time between the action starts and ends, before sending the event.

## Testing information
- In POS, add some items to cart, tap checkout, go back to the cart (make some edits or not) and tap checkout again, then follow the card payment flow. At the end, observe that the event has been tracked with elapsed times and a correct number of `checkout_tap_count`. 
```
🔵 Tracked pos_card_present_collect_payment_success, properties: [is_wpcom_store: false, milliseconds_since_card_tapped: 5625.0, plan: , site_url: https://indiemelon.mystagingwebsite.com, checkout_tap_count: 3, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, was_ecommerce_trial: false, milliseconds_since_customer_interaction_started: 27118.0, milliseconds_since_order_creation_success: 19099.0, blog_id: -1, milliseconds_since_reader_ready_to_collect_payment: 9861.0]
```
- Repeat the flow just tapping checkout once, observe the count is 1.

While I have a specific issue logged for adding more testing around tracks, in general I found very tricky to add unit tests due to how "invasive" these are in POS, and how much we'd need to mock.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Add milliseconds_since_order_creation_success

We can add this event closer to order creation if we add it inside the OrderController sync function, however this couples both objects tighter. I think it’s fine to add the track event to the aggregate model as:

- It will not track the event as a side effect of order sync, as this is only called on checkout, and we always create a new order when this happens

- It would log the start time internally for both card and cash flows, however doesn’t matter as the event is only populated and sent for card events later on, and at this point we do not know how the customer is going to pay, as happens later in the flow.

iamgabrielma committed:
add milliseconds_since_reader_ready_to_collect_payment

iamgabrielma committed:
Add milliseconds_since_card_tapped

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Add checkout_tap_count

iamgabrielma committed:
clean up

iamgabrielma commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -438,10 +448,12 @@ private extension PointOfSaleAggregateModel {
         await orderController.syncOrder(for: cart, retryHandler: { [weak self] in
             await self?.checkOut()
         })
+        collectOrderPaymentAnalyticsTracker.trackOrderCreationSuccess()
```
I had my doubts if this was the best place to add the order creation success event, I left my comments in the commit 3f9b72b9625c7b36533dd3f3211f5e4d70a83010

staskus replied:
Thanks for the commit comment. I generally agree with the reasoning. I do have a question though:
> ... and we always create a new order when this happens

Do we? On iOS we don't create a new order unless order items change. I know it's different on Android. Will these events represent the same thing due to these platform differences?

 In `PointOfSaleOrderController` we have a specific `analytics.track(.orderCreationSuccess)` event to track order creation success.

staskus replied:
If we need, one way to know if the order was actually created is to make `orderController.syncOrder` to return some `Result<SyncOrderState, Error>`. It could return an enum that tells whether the new order was created, didn't change, or error happened.

iamgabrielma replied:
Oh! Interesting, in that case using the `Result` approach feels more adequate here as we need to use the output. I've made the changes on https://github.com/woocommerce/woocommerce-ios/pull/15148/commits/0dd8512368858e46449e5ce9b4afb693845c1aab and added additional tests, so now we're more explicit about it and we do:

```
if case .success(.newOrder) = syncOrderResult {
  collectOrderPaymentAnalyticsTracker.trackOrderCreationSuccess()
}
```

iamgabrielma committed:
make test target compile

iamgabrielma committed:
Unit test for trackCheckoutTapped

iamgabrielma committed:
DI analytics and additional unit tests for POSCollectOrderPaymentAnalytics

staskus commented on WooCommerce/WooCommerceTests/POS/Analytics/POSCollectOrderPaymentAnalyticsTests.swift:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @Test func POSCollectOrderPaymentAnalyticsTests_when_successful_payment_then_tracks_event_and_properties() {
+        // Given
```
I think this test is fine 👍 

To have tests that calculate event timing logic, we would need for `POSCollectOrderPaymentAnalytics` methods or initializer to inject some date handler that would allow us to inject specific date instead of `Date()`.

iamgabrielma replied:
That sounds like a good idea, I've logged an enhancement here: https://github.com/woocommerce/woocommerce-ios/issues/15179

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -365,6 +367,14 @@ private extension PointOfSaleAggregateModel {
                 let newPaymentState = PointOfSalePaymentState(from: paymentEvent,
                                                               using: presentationStyleDeterminerDependencies)
 
+                if case .card(.acceptingCard) = newPaymentState {
```
👍 The fact that we can track all this in an aggregate model is nice. I agree that spreading it around other places of POS would become messy.

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -227,6 +227,7 @@ extension PointOfSaleAggregateModel {
         }
         do {
             try await collectPayment(for: order)
+            collectOrderPaymentAnalyticsTracker.resetCheckoutTapCountTracker()
```
Could tracker reset all properties when `trackSuccessfulPayment` is called?

iamgabrielma replied:
Good point, updated here: https://github.com/woocommerce/woocommerce-ios/pull/15148/commits/b0465370292409b0dea466bfcca0043e59496e87 . Will also move the cash bit in the next PR, as there's a "successful cash payment" function too

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -365,6 +367,14 @@ private extension PointOfSaleAggregateModel {
+                }
+
+                if case .card(.processingPayment) = newPaymentState {
+                    collectOrderPaymentAnalyticsTracker.trackCardReaderTapped()
```
One downside is that `processing` is called multiple times: 

<img width="254" alt="image" src="https://github.com/user-attachments/assets/879582e9-da5e-4aac-8ec0-ccfac7848b50" />

The difference between the first and last event was 3 seconds. I wonder what are the ways. We could only track the first event and reset it if payment is failed or successful. Do you have some ideas? Maybe Android implementation has some clues?

iamgabrielma replied:
I knew about the multiple events but I left it intentionally, as I assumed we should just capture the last event, however, if this happens seconds after the actual card tap and there's a seconds difference then is no good .

I'll check now with Android to see if it's our use of Combine or they also have to tackle the same issue.

One option to capture the first event instead would be to set another prop in the analytics tracker, sort of `hasTrackedProcessingPaymentInSession`, which also resets with the "session" after payment success:
```
if case .card(.processingPayment) = newPaymentState, !collectOrderPaymentAnalyticsTracker.hasTrackedProcessingPaymentInSession {
  collectOrderPaymentAnalyticsTracker.trackCardReaderTapped()
}
```
Otherwise could get quite tricky to filter out the specific payment event 🤔

staskus replied:
Yeah.. since we want to track "card reader tapped", the closest event is the _first_ processing event. 

I agree it's tricky. We should be careful adding `hasTrackedProcessingPaymentInSession`  since payment can fail and we may need to track "card reader tapped" for the second try.

staskus replied:
My suggestion is to take a look at additional payment events. For example,[ .cardRemoved](https://github.com/woocommerce/woocommerce-ios/blob/bfafb75aacb1354f698c6734fc292677eb2aeef2/WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift#L114). 

At least from the top, it looks closer to the intended card-tapped event compared to payment processing.

iamgabrielma replied:
Thanks for the suggestion, as discussed in slack since we cannot use `.cardRemoved` I went ahead by adding a flag internally in the analytics tracker: https://github.com/woocommerce/woocommerce-ios/pull/15148/commits/f2668a93696303df2240a576023b5a14576d86b3 this is reset as soon as the payment is successful or a new user interaction starts

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -438,10 +448,12 @@ private extension PointOfSaleAggregateModel {
 extension PointOfSaleAggregateModel {
     @MainActor
     func checkOut() async {
+        collectOrderPaymentAnalyticsTracker.trackCheckoutTapped()
```
It doesn't get reset when we close POS. Maybe it could be considered intentional but probably not. My understanding is that all the events should reset when customer interaction starts. That would be the clearest indication from which all the other events could derive. However, it needs to be synced with android.

It could reset:
- At the same time when customer interaction starts (item is added to empty cart)
- When payment finishes
- When POS launches

iamgabrielma replied:
> It doesn't get reset when we close POS

I don't think it should, but reset only after the payment finishes. I'll check this one with Android as well just to be sure we're tracking the same thing. The whole "user interaction" or "session" definition is not very solid at the moment.

iamgabrielma replied:
Ref: p1739867406250859?thread_ts=1739867347.923049&cid=C070SJRA8DP-slack-C070SJRA8DP

staskus replied:
> I don't think it should, but reset only after the payment finishes.

That's why we need to sync. For me, it makes sense to only track it in the scope of the session, which would make it reset if we detect new first user interaction. But I understand that others may think differently.

iamgabrielma replied:
Thanks! I've updated this on https://github.com/woocommerce/woocommerce-ios/pull/15148/commits/efff5ea79fa41bb0cc301b3fa31b18f4dc6a8f9e , so we reset all ongoing counters when anything considered a "user interaction starts" happens.

staskus reviewed (commented):
Thanks! It works well! However, there are a few concerns to address:

1. If we consider order creation always when we checkout or only when the new order is actually created?
2. Processing event happens multiple times, even a couple of seconds after the tap.
3. Checkout tracker is not reset if POS is closed and reopened.

iamgabrielma committed:
update syncOrder signature to result a result type

iamgabrielma committed:
move reset to analytics tracker

iamgabrielma committed:
add hasTrackedProcessingPayment flag

The Stripe SDK returns multiple `.processing` events, but we want to capture the first one in the stream only. This flag is reset as soon as the payment has been successful

iamgabrielma committed:
Reset all ongoing counters on user interaction start

iamgabrielma commented:
Thanks for the review @staskus ! This is ready for another round. Just a recap as per slack:

> If we consider order creation always when we checkout or only when the new order is actually created?

When it's actually created we trigger the event as soon as possible. At the moment we track the `orderCreationSuccess`  event in the order controller, but we milliseconds for "order success" just after we exit the function in the aggregate model. Ideally we would track those two at the same time but it's going to make everything very messy, so it doesn't seem to be worth it.

> Processing event happens multiple times, even a couple of seconds after the tap.

We resolved this by having a flag in the analytics tracker, as we couldn't rely on `.cardRemoved`. the flag resets on payment success and on user interaction starts. This also helps to reset it when there's a cash payment as a side-effect, but we can trigger it explicitly in the next PR when handling those.

> Checkout tracker is not reset if POS is closed and reopened.

That's alright, we reset trackers on user interaction start 👍

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleOrderController.swift:
```diff
@@ -51,16 +61,16 @@ protocol PointOfSaleOrderControllerProtocol {
         guard !orderState.isSyncing,
               !posCartItems.matches(order: order) else {
-            return
+            return .failure(SyncOrderStateError.syncFailure)
```
If `posCartItems.matches(order: order)` it means that that  `.success(.orderNotChanged)` not shat syncing failed.

iamgabrielma replied:
Appreciated, I got confused with the negative guard 🙇 fixed here and added a new case and test: https://github.com/woocommerce/woocommerce-ios/pull/15148/commits/af7ce8f28fa33304f031bf1553f9c0b2e403248c

staskus reviewed (approved):
Thank you for the fixes! 

Left one comment about `syncOrder` state

iamgabrielma committed:
fix issue with returning the wrong result on sync
